components/memory/watcher.go

```diff
@@ -0,0 +1,120 @@
+package memory
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/leptonai/gpud/components"
+	"github.com/leptonai/gpud/components/common"
+	events_db "github.com/leptonai/gpud/components/db"
+	memory_dmesg "github.com/leptonai/gpud/components/memory/dmesg"
+	"github.com/leptonai/gpud/log"
+	pkg_dmesg "github.com/leptonai/gpud/pkg/dmesg"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+type watcher struct {
+	ctx         context.Context
+	eventsStore events_db.Store
+
+	closeOnce    sync.Once
+	dmesgWatcher pkg_dmesg.Watcher
+}
+
+func startWatch(ctx context.Context, eventsStore events_db.Store) (*watcher, error) {
+	dw, err := pkg_dmesg.NewWatcher()
+	if err != nil {
+		return nil, err
+	}
+
+	w := &watcher{
+		ctx:          ctx,
+		eventsStore:  eventsStore,
+		dmesgWatcher: dw,
+	}
+	go w.watch()
+
+	return w, nil
+}
+
+const EventKeyLogLine = "log_line"
+
+func (w *watcher) watch() {
+	ch := w.dmesgWatcher.Watch()
+	for {
+		select {
+		case <-w.ctx.Done():
+			return
+		case line, open := <-ch:
+			if !open {
+				return
+			}
+			if line.IsEmpty() {
+				continue
+			}
+
+			ev := components.Event{
+				Time: metav1.Time{Time: line.Timestamp.UTC()},
+				Type: common.EventTypeWarning,
+				ExtraInfo: map[string]string{
+					EventKeyLogLine: line.Content,
+				},
+			}
+
+			switch {
+			case memory_dmesg.HasOOM(line.Content):
```

Comment:
I wonder if we should store a map from regex to <name, message> so we do not have to write a list of switch...case and functions like HasOOM, HasOOMKillConstraint, HasOOMKiller etc